#! /bin/sh -e

# DP: Run the ACATS testsuite using expect

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0


2004-11-28  James A. Morrison  <phython@gcc.gnu.org>

	PR ada/18302
	* ada/acats/run_all.sh (target_run): Run test through run_test.exp.
	* ada/acats/run_test.exp: Expect script for ACATS tests.

--- gcc/ada/Make-lang.in.orig	2004-11-27 07:51:09.000000000 +0100
+++ gcc/ada/Make-lang.in	2005-01-09 15:58:50.000000000 +0100
@@ -852,6 +852,7 @@
 check-gnat:
 	test -d $(ACATSDIR) || mkdir -p $(ACATSDIR)
 	testdir=`cd ${srcdir}/${ACATSDIR}; ${PWD_COMMAND}`; \
+	EXPECT=${EXPECT} ; export EXPECT ; \
 	export testdir; cd $(ACATSDIR); $${testdir}/run_acats $(CHAPTERS)
 
 .PHONY: check-gnat
diff -urN gcc/testsuite/ada/acats.orig/run_test.exp gcc/testsuite/ada/acats/run_test.exp
--- gcc/testsuite/ada/acats.orig/run_test.exp	1970-01-01 01:00:00.000000000 +0100
+++ gcc/testsuite/ada/acats/run_test.exp	2005-01-09 15:59:31.000000000 +0100
@@ -0,0 +1,13 @@
+#!/usr/bin/expect -f
+
+if {[info exists env(DEJAGNU_TIMEOUT)]} {
+  set timeout $env(DEJAGNU_TIMEOUT)
+} else {
+  set timeout 300 
+}
+
+spawn $argv
+expect timeout {
+       send_user "Program timed out.\n"
+       exit 1
+}
diff -urN gcc/testsuite/ada/acats.orig/run_all.sh gcc/testsuite/ada/acats/run_all.sh
--- gcc/testsuite/ada/acats/run_all.sh.old	2005-04-12 12:37:07.570974000 +0200
+++ gcc/testsuite/ada/acats/run_all.sh	2005-04-12 12:38:31.421765348 +0200
@@ -13,7 +13,7 @@
 gnatflags="-gnatws"
 
 target_run () {
-$*
+	$EXPECT -f $testdir/run_test.exp $1 > $2
 }
 
 # End of customization section.
@@ -92,8 +92,8 @@
 
 # Find out the size in bit of an address on the target
 target_gnatmake $testdir/support/impbit.adb >> $dir/acats.log 2>&1
-target_run $dir/support/impbit > $dir/support/impbit.out 2>&1
-target_bit=`cat $dir/support/impbit.out`
+target_run $dir/support/impbit $dir/support/impbit.out
+target_bit=`tail -1 $dir/support/impbit.out`
 echo target_bit="$target_bit" >> $dir/acats.log
 
 # Find out a suitable asm statement
@@ -259,7 +259,7 @@
       if [ ! -x $dir/tests/$chapter/$i/$binmain ]; then
          sync
       fi
-      target_run $dir/tests/$chapter/$i/$binmain > $dir/tests/$chapter/$i/${i}.log 2>&1
+      target_run $dir/tests/$chapter/$i/$binmain $dir/tests/$chapter/$i/${i}.log
       cd $dir/tests/$chapter/$i
       cat ${i}.log >> $dir/acats.log
       egrep -e '(==== |\+\+\+\+ |\!\!\!\! )' ${i}.log > /dev/null 2>&1
